harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,881 @@
+            return {}
+
+        logging.debug(f"Done training step: {num_inputs} records in {num_batches} batches.")
+        self.prof.set_training(False)
+
+        return metrics
+
+    @torch.no_grad()  # type: ignore
+    def _compute_validation_metrics(self) -> workload.Response:
+        self.context.reset_reducers()
+        # Set the behavior of certain layers (e.g., dropout) that are
+        # different between training and inference.
+        for model in self.context.models:
+            model.eval()
+
+        for callback in self.callbacks.values():
+            if util.is_overridden(callback.on_validation_step_start, pytorch.PyTorchCallback):
+                logging.warning(
+                    "on_validation_step_start is now deprecated, "
+                    "please use on_validation_start instead"
+                )
+                callback.on_validation_step_start()
+
+        for callback in self.callbacks.values():
+            callback.on_validation_start()
+
+        num_inputs = 0
+        keys = None
+        batch_metrics = []
+
+        for callback in self.callbacks.values():
+            callback.on_validation_epoch_start()
+
+        validation_iterator = iter(self.validation_loader) if self.validation_loader else None
+        for idx in range(cast(int, self.num_validation_batches)):
+            num_inputs += cast(int, self.validation_batch_size)
+            # Note that when using pipeline parallelism, each call to evaluate_batch will request
+            # self.context.num_micro_batches_per_slot batches from the validation iterator.
+            # This is why we set self.num_validation_batches differently for pipeline parallel
+            # and no pipeline parallel when building the datalaoders.
+            vld_metrics = self.trial.evaluate_batch(validation_iterator, idx)
+            if self.context.mpu.should_report_metrics():
+                # Verify validation metric names are the same across batches.
+                if keys is None:
+                    keys = vld_metrics.keys()
+                else:
+                    if keys != vld_metrics.keys():
+                        raise det.errors.InvalidExperimentException(
+                            "Validation metric names must match across all batches of data.",
+                        )
+                if not isinstance(vld_metrics, dict):
+                    raise det.errors.InvalidExperimentException(
+                        "validation_metrics() must return a "
+                        "dictionary of string names to Tensor "
+                        "metrics",
+                    )
+                # TODO: For performance perform -> cpu() only at the end of validation.
+                batch_metrics.append(_metric_utils._convert_metrics_to_numpy(vld_metrics))
+            if self.env.test_mode:
+                break
+
+        all_keys = self.context.distributed._zmq_gather(keys if keys is None else list(keys))
```

Comment:
neat trick, you can use `keys and list(keys)` here, though you'd have to also rely on if keys was truthy below (rather than an explicit none check):  

```diff
- all_keys = [k for k in all_keys if k is not None]
+ all_keys = [k for k in all_keys if k]

Reply:
:+1: